crates/connector_proxy/src/errors.rs

```diff
@@ -1,64 +1,82 @@
+use bytes::Bytes;
+use futures::{TryStream, TryStreamExt, Stream};
+use validator::ValidationErrors;
+
+use crate::apis::InterceptorStream;
+
 #[derive(thiserror::Error, Debug)]
 pub enum Error {
-    #[error("failed in starting bouncer process.")]
-    BouncerProcessStartError,
-
-    #[error("channel timeout in receiving messages after 5 seconds.")]
+    #[error("go.estuary.dev/E001: Network Tunnel startup timeout of 5 seconds exceeded. Please troubleshoot your network tunnel configuration and connection and try again")]
     ChannelTimeoutError,
 
-    #[error("command execution failed: {0}.")]
+    #[error("go.estuary.dev/E002: Failed to execute command: {0}")]
     CommandExecutionError(String),
 
-    #[error("duplicated key: {0}.")]
+    #[error("go.estuary.dev/E003: \"{0}\" key already exists in connector's endpoint specification schema, unable to add this key to the endpoint specification schema")]
     DuplicatedKeyError(&'static str),
 
-    #[error("Entrypoint is an empty string.")]
+    #[error("go.estuary.dev/E004: Unable to find the entrypoint of the connector's container. Please make sure your container defines a valid entrypoint")]
     EmptyEntrypointError,
 
-    #[error("unable to parse the inspect file.")]
+    #[error("go.estuary.dev/E005: Unable to parse the container image inspect file")]
     InvalidImageInspectFile,
 
-    #[error("missing process io pipes.")]
+    #[error("go.estuary.dev/E006: Unable to create an IO pipe to the connector")]
     MissingIOPipe,
 
-    #[error("mismatching runtime protocol")]
-    MismatchingRuntimeProtocol,
-
-    #[error("No ready signal is received. {0}")]
-    NotReady(&'static str),
-
-    #[error("invalid endpoint json config.")]
-    InvalidEndpointConfig,
+    #[error("go.estuary.dev/E007: The connector's protocol does not match the requested protocol. Connector protocol is {0}, requested protocol is {1}")]
+    MismatchingRuntimeProtocol(String, &'static str),
 
-    #[error("invalid json pointer '{0}' to config.")]
-    InvalidJsonPointer(String),
-
-    #[error(transparent)]
+    #[error("go.estuary.dev/E008: IO Error: {0}")]
```

Comment:
when you have a wrapped error don't directly format it into your error string. `anyhow` has nice chained-error formatting already where it will walk and format the causal error chain for you.

```
#[error("go.estuary.dev/E008: IO Error")]
IOError(#[from] std::io::Error),
```

(I've made this mistake myself in the past, and there are still places where we've done it that ought to be cleaned up, but here we are :shrug:).

It does mean that the top-level error presentation needs to print the formatted chained error. If you have a `tokio::main` as I believe you do here, though, IIRC you can just return anyhow::Error and it does the right thing:

```
#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
```